validators/valset/snapshot/set.go

```diff
@@ -0,0 +1,538 @@
+		getSigner:      getSigner,
+		candidates:     make([]*valset.Candidate, 0),
+		candidatesLock: sync.RWMutex{},
+		epochSize:      epochSize,
+	}
+
+	if err := set.initialize(); err != nil {
+		return nil, err
+	}
+
+	return set, nil
+}
+
+func (s *SnapshotValidatorSet) initialize() error {
+	header := s.blockchain.Header()
+	meta := s.GetSnapshotMetadata()
+
+	if header.Number == 0 {
+		// Add genesis
+		if err := s.addHeaderSnap(header); err != nil {
+			return err
+		}
+	}
+
+	// If the snapshot is not found, or the latest snapshot belongs to a previous epoch,
+	// we need to start rebuilding the snapshot from the beginning of the current epoch
+	// in order to have all the votes and validators correctly set in the snapshot,
+	// since they reset every epoch.
+
+	// Get epoch of latest header and saved metadata
+	var (
+		currentEpoch = header.Number / s.epochSize
+		metaEpoch    = meta.LastBlock / s.epochSize
+
+		snapshot = s.getSnapshot(header.Number)
+	)
+
+	if snapshot == nil || metaEpoch < currentEpoch {
+		// Restore snapshot at the beginning of the current epoch by block header
+		// if list doesn't have any snapshots to calculate snapshot for the next header
+		s.logger.Info("snapshot was not found, restore snapshot at beginning of current epoch", "current epoch", currentEpoch)
+		beginHeight := currentEpoch * s.epochSize
+		beginHeader, ok := s.blockchain.GetHeaderByNumber(beginHeight)
+
+		if !ok {
+			return fmt.Errorf("header at %d not found", beginHeight)
+		}
+
+		if err := s.addHeaderSnap(beginHeader); err != nil {
+			return err
+		}
+
+		s.store.updateLastBlock(beginHeight)
+
+		meta = s.GetSnapshotMetadata()
+	}
+
+	// Process headers if we missed some blocks in the current epoch
+	if header.Number > meta.LastBlock {
+		s.logger.Info("syncing past snapshots", "from", meta.LastBlock, "to", header.Number)
+
+		for num := meta.LastBlock + 1; num <= header.Number; num++ {
+			if num == 0 {
+				continue
+			}
+
+			header, ok := s.blockchain.GetHeaderByNumber(num)
+			if !ok {
+				return fmt.Errorf("header %d not found", num)
+			}
+
+			if err := s.ProcessHeader(header); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (s *SnapshotValidatorSet) SourceType() valset.SourceType {
+	return valset.Snapshot
+}
+
+func (s *SnapshotValidatorSet) GetValidators(height uint64) (validators.Validators, error) {
+	var snapshotHeight uint64 = 0
+	if int64(height) > 0 {
+		snapshotHeight = height - 1
+	}
+
+	snapshot := s.getSnapshot(snapshotHeight)
+	if snapshot == nil {
+		return nil, ErrSnapshotNotFound
+	}
+
+	return snapshot.Set, nil
+}
+
+func (s *SnapshotValidatorSet) GetSnapshotMetadata() *SnapshotMetadata {
+	return &SnapshotMetadata{
+		LastBlock: s.store.getLastBlock(),
+	}
+}
+
+func (s *SnapshotValidatorSet) GetSnapshots() []*Snapshot {
+	return s.store.list
+}
+
+func (s *SnapshotValidatorSet) UpdateSet(newValidators validators.Validators, from uint64) error {
+	snapshotHeight := from - 1
```

Comment:
Why?

Reply:
I'm respecting the current develop branch code. Snapshot height is defined is the biggest height of which header changed the snapshot. When using new validators at N, snapshot's height will be N-1 because this snapshot means the one applied headers to N-1.

Reply:
Please add this as a comment for this 🙏

Reply:
Added at 53a4cc01b0207b675a508ae6d9ff30ec43324bf8

Reply:
@zivkovicmilos ping